src/native/eventpipe/ds-ipc-win32.c

```diff
@@ -41,17 +41,20 @@
 #define DS_EXIT_BLOCKING_PAL_SECTION
 #endif
 
+#ifndef __cplusplus
 #undef ep_rt_object_alloc
 #define ep_rt_object_alloc(obj_type) ((obj_type *)calloc(1, sizeof(obj_type)))
 
-static
-inline
-void
-ep_rt_object_free (void *ptr)
-{
-	if (ptr)
-		free (ptr);
-}
+#undef ep_rt_object_free
+#define ep_rt_object_free(obj_ptr) do { if (obj_ptr) free (obj_ptr); } while(0)
+#else
+#include <new>
+#undef ep_rt_object_alloc
+#define ep_rt_object_alloc(obj_type) (new (std::nothrow) obj_type())
```

Comment:
CoreCLR had its own version of `nothrow` that is different from `std::nothrow`. If you are aligning it to what it used to do before, this should use the CoreCLR version of `nothrow`.

Reply:
Since these sources are part of debug PAL library it has very limited access to coreclr versions and I believe just using nothrow won't work as part of that build (think it will give compile errors).

Reply:
So trying to compile debug_pal_obj using nothrow will give you:

src\native\eventpipe\ds-ipc-win32.c(112,29): error C2065: 'nothrow': undeclared identifier

since debug_pal_obj doesn't build with the coreclr headers, so it will need to use std::nothrow since new and delete will also come directly from std.

Reply:
I do not think `delete` will come from `std`. You can only have one and only delete, and we rerouted to the coreclr-specific `delete` in coreclr.dll.

I think it may be best to stick with malloc/free for all cases.

Reply:
The ep_rt_object_alloc and ep_rt_object_array_alloc in EventPipe C library is intended to allocate structs and I was defensive when implementing it on mono using g_new0 (used quite frequently in Mono to allocate memory) and that will zero initialize, meaning that we will at least have a default state of all allocated structs in the EventPipe C library (no uninitialized members). We pay a small cost for being defensive like that, but if we would like to change that I will need to make sure we set all struct members in all places, since some depends on default values. If we think it is worth perf wise to not zero initialize I could make a pass and change that, but until that is done we need to use calloc/free to follow the same pattern.

Reply:
calloc/free is fine too.